src/pkg/cli/client/byoc/aws/byoc.go

```diff
@@ -281,40 +282,62 @@ func (b *ByocAws) findZone(ctx context.Context, domain, role string) (string, er
 		} else if err != nil {
 			return "", err
 		}
-		return zoneId, nil
+		return *zone.Id, nil
 	}
 }
 
 func (b *ByocAws) delegateSubdomain(ctx context.Context) (string, error) {
 	if b.ProjectDomain == "" {
-		return "", errors.New("custom domain not set")
+		panic("custom domain not set")
 	}
-	domain := b.ProjectDomain
+
 	cfg, err := b.driver.LoadConfig(ctx)
 	if err != nil {
 		return "", byoc.AnnotateAwsError(err)
 	}
 	r53Client := route53.NewFromConfig(cfg)
 
-	zoneId, err := aws.GetZoneIdFromDomain(ctx, domain, r53Client)
-	if errors.Is(err, aws.ErrNoZoneFound) {
-		zoneId, err = aws.CreateZone(ctx, domain, r53Client)
-		if err != nil {
-			return "", byoc.AnnotateAwsError(err)
-		}
-	} else if err != nil {
-		return "", byoc.AnnotateAwsError(err)
+	zone, err := aws.GetHostedZoneByName(ctx, b.ProjectDomain, r53Client)
+	if !errors.Is(err, aws.ErrZoneNotFound) {
+		// return "", byoc.AnnotateAwsError(err)
 	}
 
-	// Get the NS records for the subdomain zone and call DelegateSubdomainZone again
-	nsServers, err := aws.GetRecordsValue(ctx, zoneId, domain, r53types.RRTypeNs, r53Client)
+	// Get the NS records for the delegation set and let Pulumi create the hosted zone for us
+	var zoneId *string
+	if zone != nil {
+		zoneId = zone.Id
+	}
+	delegationSet, err := aws.CreateDelegationSet(ctx, zoneId, r53Client)
+	var delegationSetAlreadyCreated *r53types.DelegationSetAlreadyCreated
+	var delegationSetAlreadyReusable *r53types.DelegationSetAlreadyReusable
+	if errors.As(err, &delegationSetAlreadyCreated) || errors.As(err, &delegationSetAlreadyReusable) {
+		term.Debug("Route53 delegation set already created:", err)
+		delegationSet, err = aws.GetDelegationSet(ctx, r53Client)
+	}
 	if err != nil {
 		return "", byoc.AnnotateAwsError(err)
 	}
+	term.Debug("Route53 delegation set ID:", *delegationSet.Id)
+	b.delegationSetId = strings.TrimPrefix(*delegationSet.Id, "/delegationset/")
+	nsServers := delegationSet.NameServers
 	if len(nsServers) == 0 {
 		return "", errors.New("no NS records found for the subdomain zone")
 	}
 
+	if zone != nil {
+		// Get the NS records for the subdomain zone and call DelegateSubdomainZone again
+		nsServers2, err := aws.ListResourceRecords(ctx, *zone.Id, b.ProjectDomain, r53types.RRTypeNs, r53Client)
+		if err != nil {
+			return "", byoc.AnnotateAwsError(err)
+		}
+		// Ensure the NS records match the ones from the delegation set
+		sort.Strings(nsServers)
+		sort.Strings(nsServers2)
+		if !slices.Equal(nsServers, nsServers2) {
+			term.Warnf("NS records for the existing subdomain zone do not match the delegation set: %v <> %v", nsServers, nsServers2)
```

Comment:
Is this check necessary?
- How and how often would this happend?
- if this happens, what do we expect the user to do? Is it something that we can do for the user?